### ISSUE TITLE
[Feature] Override custom STKEYS offsets

### ISSUE THREAD

**Issue by NightFright2k19** [fabiangreffrath/woof#1747]
Some PWADs provide their own graphics for the HUD keys (STKEYSx). In at least one particular case I'm aware of, i.e. ["Judgment: Retrial"](https://www.doomworld.com/forum/topic/125238), these graphics come with strange offsets (e.g. 0/-19 or 0/19). While these don't make any particular difference ingame or in the normal statusbar, their alignments are screwed up in the Boom fullscreen HUD.

My suggestion is to ignore such offsets and use the defaults from the IWADs instead to avoid such problems. To make it fool-proof, I guess this should only be done if sprite dimensions match (i.e. if the only difference between the original STKEYS graphics and the custom ones are palette and offset adjustments).

**Comment by fabiangreffrath** [comment]
Do other ports like DSDA-Doom also face the same problem?

**Comment by NightFright2k19** [comment]
In DSDA-Doom, I cannot figure out how to invoke the classic Boom HUD with left-aligned stats. The way it is by default, keys are aligned top to bottom, which doesn't seem to be an issue.

In PrBoom+, however, keys are aligned from left to right. There it can be reproduced, as seen here:
![prboom-plus](https://github.com/fabiangreffrath/woof/assets/54737241/a6cd822d-f7bd-45ac-917a-406880047d69)

**Comment by fabiangreffrath** [comment]
I wonder how `patch->topoffset` if ignored when drawing the icons to the status bar but not when drawing them on the Boom HUD?

**Comment by NightFright2k19** [comment]
Maybe it's even a bug. TBH I really don't see any point in giving these graphics any sort of custom offsets since it is basically needless when using original graphics. I have also contacted the PWAD author for a statement.

That being said, I guess some sort of failsafe mechanism for such (rare) cases might deserve some consideration.

**Comment by fabiangreffrath** [comment]
The `patch->topoffset` values are used to switch the vertical positions of the red and blue keys on the status bar. Compare Vanilla Doom with JudgmentRT.wad:

![woof0000](https://github.com/fabiangreffrath/woof/assets/6449237/86558d33-4986-4a65-a2f8-7be1840d15f5)

![woof0001](https://github.com/fabiangreffrath/woof/assets/6449237/4fbae169-510e-473b-b76c-6c5851e53469)

And this is how it looks if I force the `patch->topoffset` values to `0`:

![woof0002](https://github.com/fabiangreffrath/woof/assets/6449237/86965363-8cf7-4872-8c08-360b067f06a4)

![woof0004](https://github.com/fabiangreffrath/woof/assets/6449237/2aef3c3e-d50c-4de6-bbce-063bf2d5aeda)

**Comment by NightFright2k19** [comment]
Interesting. It still doesn't make any sense unless the author is suffering from some sort of OCD.

Which leaves us with the question:
Should Woof! be capable of compensating such stunts when using the Boom HUD or do we have to count on the user to fix this by reversing the offsets manually?

**Comment by fabiangreffrath** [comment]
> Interesting. Apparently this switch was done so that the red keys are better visible in front of the custom statusbar. This could have been solved differently, I presume.

The obvious solution would have been to just switch out the graphics lumps for the red and blue key/skull cards.

> If the offset fix can only be done in general and not for the Boom HUD alone: From my POV, a flawless HUD display takes clear priority over some questionable icon sorting issue. Personally I wouldn't even notice this adjustment.

Our general stance is that we respect author choices. So, in this case, when the author has chosen to manipulate the vertical positions of the key graphics patches instead of merely switching them in the WAD, I'd say that it's broken by design.

You could also mess up the Boom HUD by manipulating the sizes and positions of the HUD font glyph patches. However, we don't check (and fix) these either.

My solution to this would be to include a supplemental WAD in the `autoload/JudgmentRT.wad` which contains the original key/skull graphics lumps from the IWAD.

**Comment by NightFright2k19** [comment]
Well, I guess it would open all kinds of doors for follow-up side fixes since key graphics can vary in dimensions and offsets. An external patch file is also better for EOL ports such as PrBoom+ which wouldn't receive any further updates. 

In any case, I have made a STKEYS patch for Judgment: Retrial, which I provide here. I leave it up to you whether you wish to include it in future releases (I doubt the author will). 
[judgmentrt_keyfix.zip](https://github.com/user-attachments/files/15881823/judgmentrt_keyfix.zip)

**Comment by fabiangreffrath** [comment]
Sorry, but I don't think we are going to provide fixes for random community PWADs in our source port distribution.

**Comment by NightFright2k19** [comment]
Fair enough. It would only be this one case, anyway, so its use would be very limited. Guess this means we're done here.
